File: app/api/watchlist.py
"""
Watchlist API

Enhanced watchlist system with categorized entries:
- VIP: Fast-track check-in, notify leadership
- Blacklist: Block entry, alert security
- Restricted: Extra verification required
- Banned: Permanent block, legal hold
"""
from flask import Blueprint, request, jsonify
from bson import ObjectId
from datetime import datetime, timezone
from enum import Enum

from app.db import get_db, visitor_collection
from app.auth import require_auth, require_company_access
from app.utils import get_current_utc
from app.services.audit_logger import log_action

watchlist_bp = Blueprint('watchlist', __name__)


class WatchlistCategory(str, Enum):
    VIP = "vip"                    # Fast-track, VIP treatment
    BLACKLIST = "blacklist"        # Block entry, alert security
    RESTRICTED = "restricted"      # Extra verification needed
    BANNED = "banned"              # Permanent block, legal hold
    WATCHLIST = "watchlist"        # General monitoring


def get_watchlist_collection():
    """Get the watchlist collection"""
    return get_db()['watchlist']


def convert_objectids(obj):
    """Recursively convert ObjectIds to strings"""
    if isinstance(obj, dict):
        return {k: convert_objectids(v) for k, v in obj.items()}
    elif isinstance(obj, list):
        return [convert_objectids(i) for i in obj]
    elif isinstance(obj, ObjectId):
        return str(obj)
    elif isinstance(obj, datetime):
        return obj.isoformat()
    return obj


@watchlist_bp.route('/entries', methods=['GET'])
@require_company_access
def list_watchlist_entries():
    """
    List all watchlist entries for a company.
    
    Query Parameters:
        companyId (required): Company ObjectId
        category (optional): Filter by category (vip, blacklist, restricted, banned)
        active (optional): Filter by active status (default: true)
        limit (optional): Number of records (default: 100)
    """
    try:
        company_id = request.args.get('companyId') or getattr(request, 'company_id', None)
        category = request.args.get('category')
        active = request.args.get('active', 'true').lower() == 'true'
        limit = int(request.args.get('limit', 100))
        
        if not company_id:
            return jsonify({'error': 'Company ID is required'}), 400
        
        watchlist_collection = get_watchlist_collection()
        
        query = {'companyId': company_id, 'active': active}
        if category:
            query['category'] = category
        
        entries = list(watchlist_collection.find(query).sort('createdAt', -1).limit(limit))
        
        # Count by category
        category_counts = {}
        for cat in WatchlistCategory:
            count = watchlist_collection.count_documents({
                'companyId': company_id,
                'category': cat.value,
                'active': True
            })
            category_counts[cat.value] = count
        
        return jsonify({
            'entries': convert_objectids(entries),
            'count': len(entries),
            'categoryCounts': category_counts
        }), 200
        
    except Exception as e:
        print(f"Error listing watchlist: {e}")
        import traceback
        traceback.print_exc()
        return jsonify({'error': str(e)}), 500


@watchlist_bp.route('/entries', methods=['POST'])
@require_company_access
def add_watchlist_entry():
    """
    Add a person to the watchlist.
    
    Request Body:
        companyId (required): Company ObjectId
        category (required): vip, blacklist, restricted, banned, watchlist
        name (required): Person's name
        phone (optional): Phone number
        email (optional): Email address
        organization (optional): Organization
        reason (required for blacklist/banned): Reason for listing
        visitorId (optional): Link to existing visitor record
        idType (optional): ID document type
        idNumber (optional): ID document number
        expiresAt (optional): When the entry expires
        notes (optional): Additional notes
        photo (optional): Base64 encoded photo
    """
    try:
        data = request.json or {}
        company_id = data.get('companyId') or getattr(request, 'company_id', None)
        
        if not company_id:
            return jsonify({'error': 'Company ID is required'}), 400
        
        category = data.get('category')
        if not category or category not in [c.value for c in WatchlistCategory]:
            return jsonify({'error': 'Valid category is required (vip, blacklist, restricted, banned, watchlist)'}), 400
        
        if not data.get('name'):
            return jsonify({'error': 'Name is required'}), 400
        
        if category in ['blacklist', 'banned'] and not data.get('reason'):
            return jsonify({'error': 'Reason is required for blacklist/banned entries'}), 400
        
        watchlist_collection = get_watchlist_collection()
        
        # Check if already exists with same phone/email
        if data.get('phone'):
            existing = watchlist_collection.find_one({
                'companyId': company_id,
                'phone': data['phone'],
                'active': True
            })
            if existing:
                return jsonify({
                    'error': 'Person already on watchlist',
                    'existingId': str(existing['_id']),
                    'existingCategory': existing.get('category')
                }), 409
        
        entry_doc = {
            '_id': ObjectId(),
            'companyId': company_id,
            'category': category,
            'name': data['name'],
            'phone': data.get('phone'),
            'email': data.get('email'),
            'organization': data.get('organization'),
            'reason': data.get('reason'),
            'visitorId': ObjectId(data['visitorId']) if data.get('visitorId') else None,
            'idType': data.get('idType'),
            'idNumber': data.get('idNumber'),
            'expiresAt': datetime.fromisoformat(data['expiresAt'].replace('Z', '+00:00')) if data.get('expiresAt') else None,
            'notes': data.get('notes'),
            'hasPhoto': bool(data.get('photo')),
            'active': True,
            'createdAt': get_current_utc(),
            'createdBy': getattr(request, 'user_id', 'system'),
            'lastUpdated': get_current_utc()
        }
        
        watchlist_collection.insert_one(entry_doc)
        
        # If linked to a visitor, update their status
        if data.get('visitorId') and category in ['blacklist', 'banned']:
            visitor_collection.update_one(
                {'_id': ObjectId(data['visitorId'])},
                {
                    '$set': {
                        'blacklisted': True,
                        'blacklistReason': data.get('reason'),
                        'blacklistedAt': get_current_utc(),
                        'watchlistCategory': category
                    }
                }
            )
        
        # Audit log
        log_action(
            action=f'watchlist.{category}_added',
            entity_type='watchlist',
            entity_id=entry_doc['_id'],
            company_id=company_id,
            user_id=getattr(request, 'user_id', None),
            details={
                'name': data['name'],
                'category': category,
                'reason': data.get('reason')
            },
            severity='warning' if category in ['blacklist', 'banned'] else 'info'
        )
        
        return jsonify({
            'message': f'Added to {category}',
            'entryId': str(entry_doc['_id']),
            'category': category
        }), 201
        
    except Exception as e:
        print(f"Error adding to watchlist: {e}")
        import traceback
        traceback.print_exc()
        return jsonify({'error': str(e)}), 500


@watchlist_bp.route('/entries/<entry_id>', methods=['GET'])
@require_company_access
def get_watchlist_entry(entry_id):
    """Get details of a specific watchlist entry"""
    try:
        watchlist_collection = get_watchlist_collection()
        
        entry = watchlist_collection.find_one({'_id': ObjectId(entry_id)})
        if not entry:
            return jsonify({'error': 'Entry not found'}), 404
        
        # Get linked visitor if any
        visitor = None
        if entry.get('visitorId'):
            visitor = visitor_collection.find_one({'_id': entry['visitorId']})
        
        return jsonify({
            'entry': convert_objectids(entry),
            'linkedVisitor': convert_objectids(visitor) if visitor else None
        }), 200
        
    except Exception as e:
        print(f"Error getting watchlist entry: {e}")
        return jsonify({'error': str(e)}), 500


@watchlist_bp.route('/entries/<entry_id>', methods=['PATCH'])
@require_company_access
def update_watchlist_entry(entry_id):
    """
    Update a watchlist entry.
    
    Request Body:
        category (optional): New category
        reason (optional): Updated reason
        notes (optional): Updated notes
        expiresAt (optional): New expiry date
    """
    try:
        data = request.json or {}
        watchlist_collection = get_watchlist_collection()
        
        entry = watchlist_collection.find_one({'_id': ObjectId(entry_id)})
        if not entry:
            return jsonify({'error': 'Entry not found'}), 404
        
        update_fields = {'lastUpdated': get_current_utc()}
        
        allowed_fields = ['category', 'reason', 'notes', 'name', 'phone', 'email', 'organization']
        for field in allowed_fields:
            if field in data:
                update_fields[field] = data[field]
        
        if 'expiresAt' in data:
            if data['expiresAt']:
                update_fields['expiresAt'] = datetime.fromisoformat(data['expiresAt'].replace('Z', '+00:00'))
            else:
                update_fields['expiresAt'] = None
        
        watchlist_collection.update_one(
            {'_id': ObjectId(entry_id)},
            {'$set': update_fields}
        )
        
        return jsonify({
            'message': 'Entry updated',
            'entryId': entry_id
        }), 200
        
    except Exception as e:
        print(f"Error updating watchlist entry: {e}")
        return jsonify({'error': str(e)}), 500


@watchlist_bp.route('/entries/<entry_id>', methods=['DELETE'])
@require_company_access
def remove_from_watchlist(entry_id):
    """
    Remove (deactivate) a watchlist entry.
    
    Request Body (optional):
        reason: Reason for removal
    """
    try:
        data = request.json or {}
        watchlist_collection = get_watchlist_collection()
        
        entry = watchlist_collection.find_one({'_id': ObjectId(entry_id)})
        if not entry:
            return jsonify({'error': 'Entry not found'}), 404
        
        # Soft delete - just deactivate
        watchlist_collection.update_one(
            {'_id': ObjectId(entry_id)},
            {
                '$set': {
                    'active': False,
                    'deactivatedAt': get_current_utc(),
                    'deactivatedBy': getattr(request, 'user_id', 'system'),
                    'deactivationReason': data.get('reason')
                }
            }
        )
        
        # If linked to visitor, update their status
        if entry.get('visitorId') and entry.get('category') in ['blacklist', 'banned']:
            visitor_collection.update_one(
                {'_id': entry['visitorId']},
                {
                    '$set': {
                        'blacklisted': False,
                        'unblacklistedAt': get_current_utc()
                    },
                    '$unset': {
                        'watchlistCategory': ''
                    }
                }
            )
        
        # Audit log
        log_action(
            action=f'watchlist.{entry.get("category")}_removed',
            entity_type='watchlist',
            entity_id=entry_id,
            company_id=entry.get('companyId'),
            user_id=getattr(request, 'user_id', None),
            details={
                'name': entry.get('name'),
                'category': entry.get('category'),
                'removalReason': data.get('reason')
            }
        )
        
        return jsonify({
            'message': 'Removed from watchlist',
            'entryId': entry_id
        }), 200
        
    except Exception as e:
        print(f"Error removing from watchlist: {e}")
        return jsonify({'error': str(e)}), 500


@watchlist_bp.route('/check', methods=['POST'])
@require_company_access
def check_against_watchlist():
    """
    Check if a person is on the watchlist.
    
    Used during visitor check-in to identify VIPs, blacklisted persons, etc.
    
    Request Body:
        companyId (required): Company ObjectId
        phone (optional): Phone number to check
        email (optional): Email to check
        name (optional): Name to check (fuzzy match)
        idNumber (optional): ID number to check
    
    Returns:
        Array of matching watchlist entries
    """
    try:
        data = request.json or {}
        company_id = data.get('companyId') or getattr(request, 'company_id', None)
        
        if not company_id:
            return jsonify({'error': 'Company ID is required'}), 400
        
        watchlist_collection = get_watchlist_collection()
        
        # Build OR query for various identifiers
        or_conditions = []
        
        if data.get('phone'):
            or_conditions.append({'phone': data['phone']})
        
        if data.get('email'):
            or_conditions.append({'email': data['email'].lower()})
        
        if data.get('idNumber'):
            or_conditions.append({'idNumber': data['idNumber']})
        
        if not or_conditions:
            return jsonify({'error': 'At least one identifier (phone, email, or idNumber) is required'}), 400
        
        query = {
            'companyId': company_id,
            'active': True,
            '$or': or_conditions
        }
        
        # Check for expired entries
        now = get_current_utc()
        query['$and'] = [
            {'$or': [
                {'expiresAt': None},
                {'expiresAt': {'$gt': now}}
            ]}
        ]
        
        matches = list(watchlist_collection.find(query))
        
        if matches:
            # Log the check
            for match in matches:
                if match.get('category') in ['blacklist', 'banned']:
                    log_action(
                        action='security.blacklist_match',
                        entity_type='watchlist',
                        entity_id=match['_id'],
                        company_id=company_id,
                        details={
                            'matchedName': match.get('name'),
                            'category': match.get('category'),
                            'searchPhone': data.get('phone'),
                            'searchEmail': data.get('email')
                        },
                        severity='warning'
                    )
        
        return jsonify({
            'matches': convert_objectids(matches),
            'matchCount': len(matches),
            'hasBlacklist': any(m.get('category') in ['blacklist', 'banned'] for m in matches),
            'hasVip': any(m.get('category') == 'vip' for m in matches),
            'hasRestricted': any(m.get('category') == 'restricted' for m in matches)
        }), 200
        
    except Exception as e:
        print(f"Error checking watchlist: {e}")
        import traceback
        traceback.print_exc()
        return jsonify({'error': str(e)}), 500


@watchlist_bp.route('/stats', methods=['GET'])
@require_company_access
def get_watchlist_stats():
    """Get watchlist statistics"""
    try:
        company_id = request.args.get('companyId') or getattr(request, 'company_id', None)
        
        if not company_id:
            return jsonify({'error': 'Company ID is required'}), 400
        
        watchlist_collection = get_watchlist_collection()
        
        stats = {
            'total': watchlist_collection.count_documents({'companyId': company_id, 'active': True}),
            'byCategory': {}
        }
        
        for cat in WatchlistCategory:
            stats['byCategory'][cat.value] = watchlist_collection.count_documents({
                'companyId': company_id,
                'category': cat.value,
                'active': True
            })
        
        return jsonify(stats), 200
        
    except Exception as e:
        print(f"Error getting watchlist stats: {e}")
        return jsonify({'error': str(e)}), 500

File: server/routes/watchlist.js
/**
 * Watchlist API
 * Blacklist and watchlist management
 * Matching Python app/api/watchlist.py
 */
const express = require('express');
const router = express.Router();
const { ObjectId } = require('mongodb');

const { collections, getDb } = require('../db');
const { requireCompanyAccess } = require('../middleware/auth');
const { convertObjectIds, isValidObjectId, validateRequiredFields } = require('../utils/helpers');

/**
 * GET /api/watchlist
 * List all watchlist entries for a company
 */
router.get('/', requireCompanyAccess, async (req, res, next) => {
    try {
        const companyId = req.query.companyId;
        const limit = parseInt(req.query.limit) || 100;
        const skip = parseInt(req.query.skip) || 0;

        if (!companyId) {
            return res.status(400).json({ error: 'Company ID is required.' });
        }

        const db = getDb();
        let query;
        if (isValidObjectId(companyId)) {
            query = { $or: [{ companyId: new ObjectId(companyId) }, { companyId }] };
        } else {
            query = { companyId };
        }

        const watchlist = await db.collection('watchlist')
            .find(query)
            .sort({ createdAt: -1 })
            .skip(skip)
            .limit(limit)
            .toArray();

        const total = await db.collection('watchlist').countDocuments(query);

        res.json({
            watchlist: convertObjectIds(watchlist),
            total,
            limit,
            skip
        });
    } catch (error) {
        console.error('Error listing watchlist:', error);
        next(error);
    }
});

/**
 * POST /api/watchlist
 * Add a person to watchlist
 */
router.post('/', requireCompanyAccess, async (req, res, next) => {
    try {
        const data = req.body;

        const validation = validateRequiredFields(data, ['companyId', 'name']);
        if (!validation.valid) {
            return res.status(400).json({ error: `Missing required fields: ${validation.missing.join(', ')}` });
        }

        const db = getDb();
        const watchlistDoc = {
            _id: new ObjectId(),
            companyId: isValidObjectId(data.companyId) ? new ObjectId(data.companyId) : data.companyId,
            name: data.name,
            phone: data.phone || null,
            email: data.email || null,
            idType: data.idType || null,
            idNumber: data.idNumber || null,
            reason: data.reason || 'No reason provided',
            category: data.category || 'blacklist', // blacklist, watchlist, vip, restricted
            severity: data.severity || 'medium', // low, medium, high
            entityType: data.entityType || null, // visitor, employee
            entityId: data.entityId && isValidObjectId(data.entityId) ? new ObjectId(data.entityId) : (data.entityId || null),
            statusHistory: [{ status: 'active', changedAt: new Date(), changedBy: data.addedBy || req.user?.name || 'system', reason: data.reason || '' }],
            addedBy: data.addedBy || req.user?.name || 'system',
            status: 'active',
            expiresAt: data.expiresAt ? new Date(data.expiresAt) : null,
            notes: data.notes || null,
            imageId: data.imageId || null,
            createdAt: new Date(),
            lastUpdated: new Date()
        };

        await db.collection('watchlist').insertOne(watchlistDoc);

        res.status(201).json({
            message: 'Added to watchlist successfully',
            _id: watchlistDoc._id.toString(),
            entry: convertObjectIds(watchlistDoc)
        });
    } catch (error) {
        console.error('Error adding to watchlist:', error);
        next(error);
    }
});

/**
 * GET /api/watchlist/:entry_id
 * Get single watchlist entry
 */
router.get('/:entry_id', requireCompanyAccess, async (req, res, next) => {
    try {
        const { entry_id } = req.params;

        if (!isValidObjectId(entry_id)) {
            return res.status(400).json({ error: 'Invalid entry ID format' });
        }

        const db = getDb();
        const entry = await db.collection('watchlist').findOne({ _id: new ObjectId(entry_id) });

        if (!entry) {
            return res.status(404).json({ error: 'Watchlist entry not found' });
        }

        // If linked to a visitor or employee, enrich with their data
        let linkedEntity = null;
        if (entry.entityId && entry.entityType) {
            try {
                const eid = isValidObjectId(entry.entityId) ? new ObjectId(entry.entityId) : entry.entityId;
                const col = entry.entityType === 'employee' ? 'employees' : 'visitors';
                linkedEntity = await db.collection(col).findOne({ _id: eid });
                if (linkedEntity) linkedEntity = convertObjectIds(linkedEntity);
            } catch (e) { /* ignore lookup failures */ }
        }

        res.json({ entry: convertObjectIds(entry), linkedEntity });
    } catch (error) {
        console.error('Error getting watchlist entry:', error);
        next(error);
    }
});

/**
 * PUT /api/watchlist/:entry_id
 * Update watchlist entry
 */
router.put('/:entry_id', requireCompanyAccess, async (req, res, next) => {
    try {
        const { entry_id } = req.params;
        const data = req.body;

        if (!isValidObjectId(entry_id)) {
            return res.status(400).json({ error: 'Invalid entry ID format' });
        }

        const updateFields = {};
        const allowedFields = ['name', 'phone', 'email', 'idType', 'idNumber', 'reason', 'category', 'severity', 'status', 'notes', 'expiresAt', 'entityType', 'entityId'];

        for (const field of allowedFields) {
            if (data[field] !== undefined) {
                if (field === 'expiresAt' && data[field]) {
                    updateFields[field] = new Date(data[field]);
                } else if (field === 'entityId' && data[field] && isValidObjectId(data[field])) {
                    updateFields[field] = new ObjectId(data[field]);
                } else {
                    updateFields[field] = data[field];
                }
            }
        }

        // Track status changes in history
        if (data.status) {
            const historyEntry = {
                status: data.status,
                changedAt: new Date(),
                changedBy: data.changedBy || req.user?.name || 'system',
                reason: data.statusChangeReason || ''
            };
            // Use $push for statusHistory alongside $set
            const db2 = getDb();
            await db2.collection('watchlist').updateOne(
                { _id: new ObjectId(entry_id) },
                { $push: { statusHistory: historyEntry } }
            );
        }

        if (Object.keys(updateFields).length === 0) {
            return res.status(400).json({ error: 'No fields to update' });
        }

        updateFields.lastUpdated = new Date();

        const db = getDb();
        const result = await db.collection('watchlist').updateOne(
            { _id: new ObjectId(entry_id) },
            { $set: updateFields }
        );

        if (result.matchedCount === 0) {
            return res.status(404).json({ error: 'Watchlist entry not found' });
        }

        res.json({ message: 'Watchlist entry updated successfully' });
    } catch (error) {
        console.error('Error updating watchlist:', error);
        next(error);
    }
});

/**
 * PATCH /api/watchlist/:entry_id/status
 * Quick status change (e.g. active → cleared, cleared → active)
 */
router.patch('/:entry_id/status', requireCompanyAccess, async (req, res, next) => {
    try {
        const { entry_id } = req.params;
        const { status, reason } = req.body;

        if (!isValidObjectId(entry_id)) {
            return res.status(400).json({ error: 'Invalid entry ID format' });
        }

        const validStatuses = ['active', 'cleared', 'expired'];
        if (!validStatuses.includes(status)) {
            return res.status(400).json({ error: `Invalid status. Must be one of: ${validStatuses.join(', ')}` });
        }

        const db = getDb();
        const historyEntry = {
            status,
            changedAt: new Date(),
            changedBy: req.user?.name || 'system',
            reason: reason || ''
        };

        const result = await db.collection('watchlist').updateOne(
            { _id: new ObjectId(entry_id) },
            {
                $set: { status, lastUpdated: new Date() },
                $push: { statusHistory: historyEntry }
            }
        );

        if (result.matchedCount === 0) {
            return res.status(404).json({ error: 'Watchlist entry not found' });
        }

        res.json({ message: `Status changed to ${status}`, status });
    } catch (error) {
        console.error('Error changing watchlist status:', error);
        next(error);
    }
});

/**
 * DELETE /api/watchlist/:entry_id
 * Permanently remove from watchlist
 */
router.delete('/:entry_id', requireCompanyAccess, async (req, res, next) => {
    try {
        const { entry_id } = req.params;

        if (!isValidObjectId(entry_id)) {
            return res.status(400).json({ error: 'Invalid entry ID format' });
        }

        const db = getDb();
        const result = await db.collection('watchlist').deleteOne({ _id: new ObjectId(entry_id) });

        if (result.deletedCount === 0) {
            return res.status(404).json({ error: 'Watchlist entry not found' });
        }

        res.json({ message: 'Permanently removed from watchlist' });
    } catch (error) {
        console.error('Error removing from watchlist:', error);
        next(error);
    }
});

/**
 * GET /api/watchlist/alerts
 * Escalation alerts — blacklisted/restricted persons with recent activity
 */
router.get('/alerts', requireCompanyAccess, async (req, res, next) => {
    try {
        const companyId = req.query.companyId;
        const days = parseInt(req.query.days) || 7;
        const since = new Date(Date.now() - days * 24 * 60 * 60 * 1000);

        if (!companyId) {
            return res.status(400).json({ error: 'Company ID is required' });
        }

        const db = getDb();
        const companyMatch = isValidObjectId(companyId) ? { $or: [{ companyId: new ObjectId(companyId) }, { companyId }] } : { companyId };

        // Get all active blacklisted/restricted entries
        const dangerEntries = await db.collection('watchlist').find({
            ...companyMatch,
            status: 'active',
            category: { $in: ['blacklist', 'restricted'] }
        }).toArray();

        if (dangerEntries.length === 0) {
            return res.json({ alerts: [], totalDanger: 0 });
        }

        const alerts = [];

        for (const entry of dangerEntries) {
            let recentActivity = [];

            // Build match conditions: by entityId or by name/phone
            if (entry.entityType === 'visitor' || !entry.entityType) {
                // Search visits by visitorId or name/phone
                const visitQuery = { ...companyMatch, createdAt: { $gte: since } };
                const orConds = [];
                if (entry.entityId) {
                    const eid = isValidObjectId(entry.entityId) ? new ObjectId(entry.entityId) : entry.entityId;
                    orConds.push({ visitorId: eid });
                    orConds.push({ visitorId: entry.entityId.toString() });
                }
                if (entry.name) orConds.push({ visitorName: { $regex: entry.name, $options: 'i' } });
                if (entry.phone) orConds.push({ visitorMobile: entry.phone });
                if (entry.phone) orConds.push({ phone: entry.phone });

                if (orConds.length > 0) {
                    visitQuery.$or = orConds;
                    const visits = await db.collection('visits').find(visitQuery).sort({ createdAt: -1 }).limit(3).toArray();
                    recentActivity.push(...visits.map(v => ({
                        type: 'visit',
                        date: v.createdAt || v.expectedArrival,
                        status: v.status,
                        purpose: v.purpose || '',
                        host: v.hostEmployeeName || '',
                        location: v.locationName || '',
                    })));
                }
            }

            if (entry.entityType === 'employee' || !entry.entityType) {
                // Search attendance by employeeId or name
                const attQuery = { ...companyMatch, date: { $gte: since } };
                const orConds = [];
                if (entry.entityId) {
                    const eid = isValidObjectId(entry.entityId) ? new ObjectId(entry.entityId) : entry.entityId;
                    orConds.push({ employeeId: eid });
                    orConds.push({ employeeId: entry.entityId.toString() });
                }
                if (entry.name) orConds.push({ employeeName: { $regex: entry.name, $options: 'i' } });

                if (orConds.length > 0) {
                    attQuery.$or = orConds;
                    const att = await db.collection('attendance').find(attQuery).sort({ date: -1 }).limit(3).toArray();
                    recentActivity.push(...att.map(a => ({
                        type: 'attendance',
                        date: a.date || a.checkIn,
                        status: a.checkOut ? 'checked-out' : 'checked-in',
                        checkIn: a.checkIn,
                        checkOut: a.checkOut,
                        source: a.source || '',
                        deviceId: a.deviceId || '',
                    })));
                }
            }

            if (recentActivity.length > 0) {
                // Sort by date descending
                recentActivity.sort((a, b) => new Date(b.date) - new Date(a.date));
                alerts.push({
                    entryId: entry._id.toString(),
                    name: entry.name,
                    phone: entry.phone,
                    category: entry.category,
                    severity: entry.severity,
                    reason: entry.reason,
                    entityType: entry.entityType || 'unknown',
                    lastSeen: recentActivity[0].date,
                    lastSeenType: recentActivity[0].type,
                    activityCount: recentActivity.length,
                    recentActivity: recentActivity.slice(0, 3),
                });
            }
        }

        // Sort alerts by most recent first
        alerts.sort((a, b) => new Date(b.lastSeen) - new Date(a.lastSeen));

        res.json({
            alerts,
            totalDanger: dangerEntries.length,
            alertCount: alerts.length,
            periodDays: days,
        });
    } catch (error) {
        console.error('Error fetching watchlist alerts:', error);
        next(error);
    }
});

/**
 * GET /api/watchlist/:entry_id/activity
 * Get recent activity (visits + attendance) for a watchlist entry
 */
router.get('/:entry_id/activity', requireCompanyAccess, async (req, res, next) => {
    try {
        const { entry_id } = req.params;
        const limit = parseInt(req.query.limit) || 10;

        if (!isValidObjectId(entry_id)) {
            return res.status(400).json({ error: 'Invalid entry ID format' });
        }

        const db = getDb();
        const entry = await db.collection('watchlist').findOne({ _id: new ObjectId(entry_id) });

        if (!entry) {
            return res.status(404).json({ error: 'Watchlist entry not found' });
        }

        const companyMatch = entry.companyId
            ? (isValidObjectId(entry.companyId) ? { $or: [{ companyId: new ObjectId(entry.companyId) }, { companyId: entry.companyId.toString() }] } : { companyId: entry.companyId })
            : {};

        const activity = [];

        // ─── Search Visits ────────────────────────────────────────
        const visitOrConds = [];
        if (entry.entityId && entry.entityType === 'visitor') {
            const eid = isValidObjectId(entry.entityId) ? new ObjectId(entry.entityId) : entry.entityId;
            visitOrConds.push({ visitorId: eid });
            visitOrConds.push({ visitorId: entry.entityId.toString() });
        }
        if (entry.name) visitOrConds.push({ visitorName: { $regex: entry.name, $options: 'i' } });
        if (entry.phone) visitOrConds.push({ visitorMobile: entry.phone });
        if (entry.phone) visitOrConds.push({ phone: entry.phone });

        if (visitOrConds.length > 0) {
            const visits = await db.collection('visits').find({
                ...companyMatch,
                $or: visitOrConds,
            }).sort({ createdAt: -1 }).limit(limit).toArray();

            activity.push(...visits.map(v => ({
                _id: v._id.toString(),
                type: 'visit',
                date: v.createdAt || v.expectedArrival,
                status: v.status || 'unknown',
                purpose: v.purpose || '',
                host: v.hostEmployeeName || '',
                location: v.locationName || '',
                checkIn: v.actualArrival || null,
                checkOut: v.actualDeparture || null,
                visitorName: v.visitorName || '',
            })));
        }

        // ─── Search Attendance ────────────────────────────────────
        const attOrConds = [];
        if (entry.entityId && entry.entityType === 'employee') {
            const eid = isValidObjectId(entry.entityId) ? new ObjectId(entry.entityId) : entry.entityId;
            attOrConds.push({ employeeId: eid });
            attOrConds.push({ employeeId: entry.entityId.toString() });
        }
        if (entry.name) attOrConds.push({ employeeName: { $regex: entry.name, $options: 'i' } });

        if (attOrConds.length > 0) {
            const att = await db.collection('attendance').find({
                ...companyMatch,
                $or: attOrConds,
            }).sort({ date: -1 }).limit(limit).toArray();

            activity.push(...att.map(a => ({
                _id: a._id.toString(),
                type: 'attendance',
                date: a.date || a.checkIn,
                status: a.checkOut ? 'checked-out' : 'checked-in',
                checkIn: a.checkIn || null,
                checkOut: a.checkOut || null,
                hoursWorked: a.hoursWorked || null,
                source: a.source || '',
                deviceId: a.deviceId || '',
                employeeName: a.employeeName || '',
            })));
        }

        // Sort all activity by date descending and take top N
        activity.sort((a, b) => new Date(b.date) - new Date(a.date));
        const trimmed = activity.slice(0, limit);

        res.json({
            activity: trimmed,
            total: trimmed.length,
            entityType: entry.entityType || 'unknown',
            name: entry.name,
        });
    } catch (error) {
        console.error('Error fetching watchlist activity:', error);
        next(error);
    }
});

/**
 * POST /api/watchlist/check
 * Check if a person is on watchlist
 */
router.post('/check', requireCompanyAccess, async (req, res, next) => {
    try {
        const { companyId, phone, email, name } = req.body;

        if (!companyId) {
            return res.status(400).json({ error: 'Company ID is required' });
        }

        const db = getDb();
        const companyQuery = isValidObjectId(companyId) ? new ObjectId(companyId) : companyId;

        const orConditions = [];
        if (phone) orConditions.push({ phone });
        if (email) orConditions.push({ email });
        if (name) orConditions.push({ name: { $regex: name, $options: 'i' } });

        if (orConditions.length === 0) {
            return res.status(400).json({ error: 'At least one of phone, email, or name is required' });
        }

        const match = await db.collection('watchlist').findOne({
            companyId: companyQuery,
            status: 'active',
            $or: orConditions
        });

        res.json({
            onWatchlist: Boolean(match),
            entry: match ? convertObjectIds(match) : null
        });
    } catch (error) {
        console.error('Error checking watchlist:', error);
        next(error);
    }
});

module.exports = router;
